File: apps/collector/src/main/java/com/hes/collector/dlms/ObisCode.java
package com.hes.collector.dlms;

import lombok.Getter;
import java.util.Arrays;
import java.util.regex.Pattern;

@Getter
public class ObisCode {
    private static final Pattern OBIS_PATTERN = Pattern.compile("^\\d+\\.\\d+\\.\\d+\\.\\d+\\.\\d+\\.\\d+$");
    private final byte[] code;

    private ObisCode(byte[] code) {
        if (code == null || code.length != 6) {
            throw new IllegalArgumentException("OBIS code must be 6 bytes");
        }
        this.code = Arrays.copyOf(code, 6);
    }

    public static ObisCode parse(String obisString) {
        if (obisString == null || !OBIS_PATTERN.matcher(obisString).matches()) {
            throw new IllegalArgumentException("Invalid OBIS code format");
        }

        String[] parts = obisString.split("\\.");
        byte[] code = new byte[6];
        
        for (int i = 0; i < 6; i++) {
            int value = Integer.parseInt(parts[i]);
            if (value < 0 || value > 255) {
                throw new IllegalArgumentException("OBIS code values must be between 0 and 255");
            }
            code[i] = (byte) value;
        }

        return new ObisCode(code);
    }

    public static ObisCode of(int a, int b, int c, int d, int e, int f) {
        byte[] code = new byte[] {
            validateByte(a),
            validateByte(b),
            validateByte(c),
            validateByte(d),
            validateByte(e),
            validateByte(f)
        };
        return new ObisCode(code);
    }

    private static byte validateByte(int value) {
        if (value < 0 || value > 255) {
            throw new IllegalArgumentException("OBIS code values must be between 0 and 255");
        }
        return (byte) value;
    }

    @Override
    public String toString() {
        return String.format("%d.%d.%d.%d.%d.%d",
            code[0] & 0xFF,
            code[1] & 0xFF,
            code[2] & 0xFF,
            code[3] & 0xFF,
            code[4] & 0xFF,
            code[5] & 0xFF);
    }

    @Override
    public boolean equals(Object obj) {
        if (this == obj) return true;
        if (obj == null || getClass() != obj.getClass()) return false;
        ObisCode other = (ObisCode) obj;
        return Arrays.equals(code, other.code);
    }

    @Override
    public int hashCode() {
        return Arrays.hashCode(code);
    }
} 

File: apps/gui/src/dlms-simulator/core/dlms/ObisCode.ts
export class ObisCode {
    constructor(
        public readonly groupA: number,
        public readonly groupB: number,
        public readonly groupC: number,
        public readonly groupD: number,
        public readonly groupE: number,
        public readonly groupF: number
    ) {
        this.validate();
    }

    private validate(): void {
        if (this.groupA < 0 || this.groupA > 255) throw new Error('Group A must be between 0 and 255');
        if (this.groupB < 0 || this.groupB > 255) throw new Error('Group B must be between 0 and 255');
        if (this.groupC < 0 || this.groupC > 255) throw new Error('Group C must be between 0 and 255');
        if (this.groupD < 0 || this.groupD > 255) throw new Error('Group D must be between 0 and 255');
        if (this.groupE < 0 || this.groupE > 255) throw new Error('Group E must be between 0 and 255');
        if (this.groupF < 0 || this.groupF > 255) throw new Error('Group F must be between 0 and 255');
    }

    public toString(): string {
        return `${this.groupA}.${this.groupB}.${this.groupC}.${this.groupD}.${this.groupE}.${this.groupF}`;
    }

    public equals(other: ObisCode): boolean {
        return this.groupA === other.groupA &&
               this.groupB === other.groupB &&
               this.groupC === other.groupC &&
               this.groupD === other.groupD &&
               this.groupE === other.groupE &&
               this.groupF === other.groupF;
    }

    public static fromString(str: string): ObisCode {
        const parts = str.split('.').map(Number);
        if (parts.length !== 6) {
            throw new Error('Invalid OBIS code format');
        }
        return new ObisCode(parts[0], parts[1], parts[2], parts[3], parts[4], parts[5]);
    }
} 